Convert movement packets

Ciel1996 opened the pull request:
## PR Checklist
<!-- Insert "x" inside square brackets to check item. -->

Please check if your PR fulfills the following requirements:

- [x] All new classes have class-level documentation comments, if there are any at all
- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:


## What is the current behaviour?
The movement packets are currently actively polled and not rate limited.


## What is the new behaviour?
Resolves #468
Subscribe `MovementHandler` to `MessageBroker<Movement>` events which then get handled instead of being polled.
In order to implement the rate-limit of 30 messages per second per agent, a Polly RateLimit-Policy has been introduced.
The desired amount of packets and the timespan can be adjusted via constant/static values at the top of the class.
Each AgentId is assigned their own policy, reason is given via code comment.

## Other information
Writing a test proved difficult, as there is no way of getting a valid and initialized instance of `Agent` as far as I've seen. If there is a known way for this, please feel free to tag me and I will include tests.

Ciel1996 committed:
Initial look at movement packet

Ciel1996 committed:
- Added Movement event and subclasses with ToMovementPacket() - method
- Started to replace polling mechanism with event based mechanism

Ciel1996 committed:
- Using IMemoryCache with a TTL of 1s to rate limit packets sent

Ciel1996 commented:
@garrettluskey I'm not that happy with the `IMemoryCache` as means of rate-limiting. Do you have any experience with this? I'm only familiar with Polly's rate limit, but I'm not sure if this could be appropriate here. 

I'm open for suggestions, maybe I will try something else until I find a better solution.

Ciel1996 committed:
Removed unused using

Ciel1996 committed:
- Replaced IMemoryCache with Policy.RateLimit

garrettluskey commented on source/Missions/Services/Agents/Packets/AgentData.cs:
```diff
@@ -35,6 +35,40 @@ public AgentData(Agent agent)
+            MountData = agentMountData;
+        }
+
+        public AgentData(Agent agent, Vec2 movementDirection) : 
```
For this we want any changed data to accumulate before sending the agent data. This currently only sends 1 piece at a time.

For example
If lookDirection and movementDirection change before the tick rate we want those in the same packet.

garrettluskey commented on source/Missions/Services/Agents/Messages/MovementInputVectorChanged.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// An <see cref="IEvent"/> propagating that the <see cref="Agent"/>'s input vector has changed.
+    /// </summary>
+    public sealed class MovementInputVectorChanged : Movement
+    {
+        /// <summary>
+        /// The changed input vector.
+        /// </summary>
+        public Vec2 InputVector { get; }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="agent"></param>
+        public MovementInputVectorChanged(Agent agent, Guid guid) : base(agent, guid)
+        {
+            InputVector = agent.MovementInputVector;
+        }
+
+        /// <inheritdoc />
+        public override MovementPacket ToMovementPacket()
+        {
+            return new MovementPacket(Guid, new AgentData(Agent, InputVector));
+        }
+    }
```
```suggestion
    internal readonly struct MovementInputVectorChanged : MovementBase
    {
        /// <summary>
        /// The changed input vector.
        /// </summary>
        public Vec2 InputVector { get; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="agent"></param>
        public MovementInputVectorChanged(Agent agent) : base(agent)
        {
            InputVector = agent.MovementInputVector;
        }
    }
```

garrettluskey commented on source/Missions/Services/Agents/Messages/Movement.cs:
```diff
@@ -0,0 +1,39 @@
+    /// <summary>
+    /// An <see cref="IEvent"/> used to propagate movement related changes of an <see cref="Agent"/>.
+    /// </summary>
+    public abstract class Movement : IEvent
```
```suggestion
    internal abstract class MovementBase : IEvent
```

garrettluskey commented on source/Missions/Services/Agents/Messages/Movement.cs:
```diff
@@ -0,0 +1,39 @@
+        /// </summary>
+        public Agent Agent { get; }
+
+        /// <summary>
+        /// The <paramref name="Agent"/>'s <see cref="Guid"/>.
+        /// </summary>
+        public Guid Guid { get; }
```
```suggestion
```

garrettluskey commented on source/Missions/Services/Agents/Messages/Movement.cs:
```diff
@@ -0,0 +1,39 @@
+        /// Constructor
+        /// </summary>
+        /// <param name="agent"></param>
+        public Movement(Agent agent, Guid guid)
+        {
+            Agent = agent;
+            Guid = guid;
+        }
```
```suggestion
        public Movement(Agent agent)
        {
            Agent = agent;
        }
```

garrettluskey commented on source/Missions/Services/Agents/Packets/MovementHandler.cs:
```diff
@@ -93,23 +101,51 @@ Mission CurrentMission
         }
 
-        private async void PollAgents()
+        private void Handle_Movement(MessagePayload<Movement> payload)
```
Create an individual handler for each movement type.

Create a new class for managing movement deltas and possible publishing packets from there (if that class gets too large feel free to separate it)

garrettluskey commented on source/Missions/Services/Agents/Messages/MountDataChanged.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// An <see cref="IEvent"/> used to propagate movement related changes of <see cref="AgentMountData"/> of an <see cref="Agent"/>.
+    /// </summary>
+    public sealed class MountDataChanged : Movement
```
Update similar to MovementInputVectorChanged

garrettluskey commented on source/Missions/Services/Agents/Messages/LookDirectionChanged.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// An <see cref="IEvent"/> propagating the <see cref="Agent"/>'s change in look direction.
+    /// </summary>
+    public sealed class LookDirectionChanged : Movement
```
Update similar to MovementInputVectorChanged

garrettluskey commented on source/Missions/Services/Agents/Messages/ActionDataChanged.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// An <see cref="IEvent"/> used to propagate movement related changes of <see cref="AgentActionData"/> of an <see cref="Agent"/>.
+    /// </summary>
+    public sealed class ActionDataChanged : Movement
```
Update similar to MovementInputVectorChanged

garrettluskey commented:
Also could you create some tests for rate limiting/testing each event

Ciel1996 commented:
Thanks for the feedback. Sorry for the delay so far, gotta invest some more time into this.

Ciel1996 committed:
Reworking the message-to-packet logic

Ciel1996 committed:
- Using Timer to send packets regularly
- Unsure what to make of Polly with this, as it seems like the cleaner
  solution
- Implemented basic delta-management-logic

Ciel1996 committed:
- removed Polly
- refactored classes
- added missing comments

garrettluskey commented on source/Missions/Services/Agents/Messages/IMovement.cs:
```diff
@@ -0,0 +1,22 @@
+    /// <summary>
+    /// An <see cref="IEvent"/> used to propagate movement related changes of an <see cref="Agent"/>.
+    /// </summary>
+    interface IMovement : IEvent
```
Rename to `IMovementEvent`

garrettluskey commented on source/Missions/Services/Agents/Packets/MovementHandler.cs:
```diff
@@ -131,13 +193,18 @@ public void Handle_PeerDisconnect(MessagePayload<PeerDisconnected> payload)
             if (_agentRegistry.OtherAgents.TryGetValue(peer, out AgentGroupController controller))
             {
-                foreach (var agent in controller.ControlledAgents.Values)
+                foreach (var kv in controller.ControlledAgents)
                 {
+                    var agent = kv.Value;
+                    var guid = kv.Key;
```
```suggestion
                foreach (var guid in controller.ControlledAgents.Keys)
                {
                    var guid = kv.Key;
```

Ciel1996 replied:
could you please explain why you suggest this? I don't seem to understand. :)

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovementDelta.cs:
```diff
@@ -0,0 +1,101 @@
+        /// <param name="change"></param>
+        public void CalculateMovement(LookDirectionChanged change)
+        {
+            _lookDirectionDelta += change.LookDirection;
```
Why is this a `+=`. I think this should be only `=` unless I'm missing something

garrettluskey commented on source/Missions/Services/Agents/Messages/IMovement.cs:
```diff
@@ -0,0 +1,22 @@
+        /// <summary>
+        /// The <see cref="Agents"/>'s network <see cref="Guid"/>.
+        /// </summary>
+        Guid Guid { get; }
```
Remove Guid this should be resolved in the consumer, i.e. MovementHandler

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovementDelta.cs:
```diff
@@ -0,0 +1,101 @@
+        /// <param name="change"></param>
+        public void CalculateMovement(MovementInputVectorChanged change)
+        {
+            _inputVectorDelta += change.InputVector;
```
Why is this a `+=`. I think this should be only `=` unless I'm missing something

Ciel1996 replied:
I thought because the vectors could change multiple time before the delta would be sent to the clients. But we could give it a try using only `=`

Ciel1996 committed:
Refactored according to feedback

Ciel1996 committed:
forgot to save Missions.csproj

garrettluskey commented on source/Missions/Services/Agents/Packets/MovementHandler.cs:
```diff
@@ -93,22 +114,68 @@ Mission CurrentMission
-                        _client.SendAll(packet);
-                    }
-                }
+                throw new ArgumentException($"No {nameof(Agent)} found");
```
Have this be logged rather than an exception
```suggestion
                Logger.Error("No {agent} found", nameof(Agent));
                return;
```

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovementDelta.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <summary>
+    /// Represents the delta of an agent's movement between the time it was created and the time it was last updated.
+    /// </summary>
+    internal sealed class AgentMovementDelta
```
Make this into a `ProtoContract`. For an example see `AgentActionData`.

What's the methodology of making this class sealed?

garrettluskey replied:
Also write a serialization test for this similar to `NetworkMessageSerializationTest`

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovementDelta.cs:
```diff
@@ -0,0 +1,101 @@
+        /// <summary>
+        /// Guid of the associated <see cref="Agent"/>.
+        /// </summary>
+        public Guid Guid { get; }
```
```suggestion
        public Guid AgentId { get; }
```

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovementDelta.cs:
```diff
@@ -0,0 +1,101 @@
+        /// </summary>
+        public Guid Guid { get; }
+
+        /// <summary>
+        /// The <see cref="Agent"/> whose movement is about to be updated.
+        /// </summary>
+        public Agent Agent { get; }
```
Since this will be sent over the network Agent isn't needed.
```suggestion

```

Ciel1996 replied:
I will remove Agent and in it's stead will add Vec2 _agentMovementDirection and AgentEquipmentData as they are required for `AgentData`

Ciel1996 committed:
Made ProtoBuf compatible; Introduced test to proof ProtoBuf serialization and deserialization work

EgardA commented:
If you could make the MovementHandler use the new methods that have been created in the NetworkAgentRegistry that would be great :)

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

Ciel1996 committed:
Fixing lost updates from merge

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

Ciel1996 committed:
Merge branch 'convert-movement-packets' of https://github.com/Bannerlord-Coop-Team/BannerlordCoop into convert-movement-packets

garrettluskey commented on source/ClientDebug/ClientDebug.csproj.user:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|AnyCPU'">
     <StartArguments>/singleplayer /client _MODULES_%2aNative%2aSandBoxCore%2aSandBox%2aStoryMode%2aMissionTestMod%2a_MODULES_</StartArguments>
-    <StartProgram>..\mb2\bin\Win64_Shipping_Client\Bannerlord.exe</StartProgram>
+    <StartProgram>E:\Repos\BannerlordCoop\mb2\bin\Win64_Shipping_Client\Bannerlord.exe</StartProgram>
```
Change this back to relative path

Ciel1996 replied:
Sure… don‘t know how that got in there

garrettluskey commented on source/MissionTests/RunTimeHelper.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace IntroductionServerTests
+{
+    public sealed class RunTimeHelper
```
I don't see where this is being used. Please remove or make use of this class

Ciel1996 replied:
I removed it yesterday 🤔 will have a look again later

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovementDelta.cs:
```diff
@@ -0,0 +1,165 @@
+    /// Represents the delta of an <see cref="Agent"/>'s movement between the time it was created and the time it was last updated.
+    /// </summary>
+    [ProtoContract(SkipConstructor = true)]
+    public class AgentMovementDelta
```
This class is meant to replace the `MovementPacket` class. If there is some reason this isn't easily completable let me know.

Make this class inherit from IPacket

garrettluskey commented on source/Missions/Services/Agents/Packets/MovementHandler.cs:
```diff
@@ -35,6 +36,12 @@ public MovementPacket(Guid agentGuid, Agent agent)
             Agent = new AgentData(agent);
         }
 
+        public MovementPacket(Guid agentGuid, AgentData agentData)
```
Replace movement packet with movement delta class. Also rename "AgentMovementDelta" to "AgentMovement" as it really doesn't contain deltas rather only actual values of data that has changed

Ciel1996 committed:
Fixed some more issues found in review and replaced MovementPacket with AgentMovement

Ciel1996 committed:
Update ClientDebug.csproj.user

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovement.cs:
```diff
@@ -0,0 +1,185 @@
+        /// </summary>
+        /// <param name="agent"></param>
+        /// <param name="guid"></param>
+        public AgentMovement(
```
All of these values should initialize to default until an event happens for a given field. Create a constructor using only Guid

garrettluskey commented on source/Missions/Services/Agents/Packets/AgentData.cs:
```diff
@@ -35,6 +35,24 @@ public AgentData(Agent agent)
+            MountData = agentMountData;
+        }
+
         public void Apply(Agent agent)
```
Remove this class entirely and move the functionality here to the `AgentMovement` class

garrettluskey replied:
let me know if there is a reason this class should still exist, I could be wrong here

Ciel1996 replied:
removed it, functionality should be added to AgentMovement struct.

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovement.cs:
```diff
@@ -0,0 +1,185 @@
+        /// </summary>
+        /// <param name="agent"></param>
+        /// <returns></returns>
+        public AgentData Apply(Agent agent)
```
Move Apply functionality from AgentData to this class. Also don't apply data for default values.

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovement.cs:
```diff
@@ -0,0 +1,185 @@
+    /// Represents the delta of an <see cref="Agent"/>'s movement between the time it was created and the time it was last updated.
+    /// </summary>
+    [ProtoContract(SkipConstructor = true)]
+    public class AgentMovement : IPacket
```
Convert this to a `public readonly struct` if possible

Ciel1996 replied:
I converted it to a struct. Can't make it readonly because of the `CalculateMovement` methods which do write to them.

Ciel1996 committed:
- Removed AgentData
- Refactored AgentMovement.Apply()
- Refactored AgentMovement into struct
- Added operator overload to AgentEquipmentData due to default
  comparison in AgentMovement

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

garrettluskey committed:
Removed unused code

garrettluskey committed:
removed unused code

garrettluskey commented:
To be able to close this PR a publisher of these events is still needed.

Create a publisher class under the Agent folder that only manages the PlayerControlled Agent (Agent.Main).

In this class poll the agents values that are in AgentMovement. Save those values in a AgentMovement variable. If these values change publish an event associated with that change.

Ciel1996 committed:
- Added AgentPublisher polling Agent.Main regularly for movement changes
  and sending events down the messageBroker
- Some refactoring of events

garrettluskey committed:
Merge branch 'convert-movement-packets' of https://github.com/Bannerlord-Coop-Team/BannerlordCoop into convert-movement-packets

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

Ciel1996 commented:
Should be good to go

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

garrettluskey committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentMovement.cs:
```diff
@@ -0,0 +1,275 @@
+using TaleWorlds.Library;
+using TaleWorlds.MountAndBlade;
+
+namespace Missions.Services.Agents.Messages
```
This should exist under the Packets folder instead

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentPublisher.cs:
```diff
@@ -0,0 +1,132 @@
+using System.Threading.Tasks;
+using TaleWorlds.MountAndBlade;
+
+namespace Missions.Services.Agents.Messages
```
This should exist in `Missions.Services.Agents`, or one level up in the directory

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentPublisher.cs:
```diff
@@ -0,0 +1,132 @@
+        public AgentPublisher(IMessageBroker messageBroker, int packetUpdateRate)
+        {
+            _agent = Agent.Main;
+            if (NetworkAgentRegistry.Instance.TryGetAgentId(_agent, out _agentId))
```
Right now this if statement will enter only if the main agent is registered. Invert this to get desired functionality.
```suggestion
            if (NetworkAgentRegistry.Instance.TryGetAgentId(_agent, out _agentId) == false)
```

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentPublisher.cs:
```diff
@@ -0,0 +1,132 @@
+                _agentId = Guid.NewGuid();
+
+                // TODO: after merge this must be changed to RegisterPlayerAgent
+                NetworkAgentRegistry.Instance.RegisterControlledAgent(_agentId, _agent);
```
Use the DI agent registry
```suggestion
                _networkAgentRegistry.RegisterControlledAgent(_agentId, _agent);
```

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentPublisher.cs:
```diff
@@ -0,0 +1,132 @@
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public AgentPublisher(IMessageBroker messageBroker, int packetUpdateRate)
```
add `INetworkAgentRegistry` as a DI parameter here. Also add this class to the DI framework in `MissionModule`

garrettluskey commented on source/Missions/Services/Agents/Messages/AgentPublisher.cs:
```diff
@@ -0,0 +1,132 @@
+        {
+            var lookDirection = new LookDirectionChanged(_agent);
+
+            if (lookDirection.LookDirection != _agent.LookDirection)
```
This only does the player agent. We also want all controlled agents. To iterate through those use `_networkAgentRegistry.ControlledAgents`

garrettluskey commented on source/Missions/Services/Agents/Packets/MovementHandler.cs:
```diff
@@ -81,50 +67,91 @@ public void Dispose()
+                Logger.Error("No {agent} found", nameof(Agent));
+            }
+
+            if (_agentMovementDeltas.TryGetValue(payloadGuid, out var delta))
```
I could be wrong but if we have an older delta stored here we would use that one instead of the newer one with this logic. It should be we always use the newer one.

Ciel1996 replied:
but at this point we are only getting the deltas that will be used as a base for further calculations.
They will then all be enumerated in PopAllDeltas() and the Dictionary cleared. But I'm also not that happy with a dictionary which basically turned into a queue. Maybe I could change this, too.

Ciel1996 committed:
Fixed some namespaces

garrettluskey commented on source/Missions/Services/Agents/AgentPublisher.cs:
```diff
@@ -0,0 +1,134 @@
+        {
+            while (!_agentPollingTaskCancellationTokenSource.IsCancellationRequested && Mission.Current != null)
+            {
+                // TODO: also add all player agents
+                var agents = _networkAgentRegistry.ControlledAgents.Values.ToList();
```
Player agent will be in the controlled agents, also this does not need to be casted to a list

garrettluskey commented on source/Missions/Services/Agents/AgentPublisher.cs:
```diff
@@ -0,0 +1,134 @@
+                }
+
+                // wait a bit and then start over
+                await Task.Delay(_packetUpdateRate / 3);
```
Why `/ 3`?

Ciel1996 replied:
I just thought of a small prime number to split the update rate of packaged into smaller bits. But we can also just use _packetUpdateRate. Would be better to be honest :)

garrettluskey commented on source/Missions/Services/Agents/AgentPublisher.cs:
```diff
@@ -0,0 +1,134 @@
+            _agentPollingTask.Wait();
+        }
+
+        private async Task PollAndUpdateAgentMovement()
```
Instead of making this async, I think we should use `System.Threading.Timer` instead. Let me know what you think

Ciel1996 replied:
yeah, `Timer` would probably better because of the reduced overhead compared to async Task, will adjust it :)

Ciel1996 committed:
replaced async method by Timer as suggested;
Also removed / 3 division

garrettluskey commented on source/Missions/Services/Agents/AgentPublisher.cs:
```diff
@@ -0,0 +1,125 @@
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public AgentPublisher(IMessageBroker messageBroker, int packetUpdateRate, INetworkAgentRegistry networkAgentRegistry)
```
I think it would be easier to pass `packetUpdateRate` in a config object.

Create a new interface and class called `AgentPublisherConfig` with packetUpdateRate as a property

Ciel1996 replied:
I agree, this object/interface would make future extensions easier :)

garrettluskey commented on source/Missions/Services/Agents/AgentPublisher.cs:
```diff
@@ -0,0 +1,125 @@
+
+        private void PollAndUpdateAgentMovement(object obj)
+        {
+            // TODO: also add all player agents
```
This todo can be removed

garrettluskey reviewed (changes requested):
2 very small things to fix. Also check that this works in game. If needed I can help you get it working with some pair programming in the future. Just let me know

Ciel1996 committed:
added AgentPublisherConfig

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets

Ciel1996 committed:
Used DI for relevant classes and interfaces

Ciel1996 committed:
Fixed dependencies

Ciel1996 committed:
Movements are received by the message broker now.

Ciel1996 committed:
Merge branch 'arena-vertical-slice' into convert-movement-packets
